### PR TITLE
sentinel-dashboard add ParamFlowRuleV2 Controller(It can adapted nacos or apollo or zookeeper DataSource)

### PR DESCRIPTION
**Pull request by threedr3am**
### Describe what this PR does / why we need it

It makes it easier for users to extend the console, enabling hotspot parameter flow rule limit support using zookeeper data sources(test module has example)

### Describe how you did it
I add a ParamFlowRuleControllerV2,it can autowired test module ParamFlowRuleZookeeperProvider and ParamFlowRuleZookeeperPublisher so that it supports the use of zookeeper data source.

### Describe how to verify it
About controller,you can use html setting and echo the ParamFlowRule to verify,About  zookeeper data source privider and publisher,you can debug the zookeeper data source to see if its data can be updated and loaded.

### Special notes for reviews
I think asynchronize need't to parse results.

### CODE REVIEW

**Comment by sczyh30** [comment]
👍

File: sentinel-dashboard/src/main/webapp/resources/app/scripts/controllers/identity.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 var app = angular.module('sentinelDashboardApp');
 
 app.controller('IdentityCtl', ['$scope', '$stateParams', 'IdentityService',
-  'ngDialog', 'FlowServiceV1', 'DegradeService', 'AuthorityRuleService', 'ParamFlowService', 'MachineService',
+  'ngDialog', 'FlowServiceV1', 'DegradeService', 'AuthorityRuleService', 'ParamFlowServiceV2', 'MachineService',
```
Comment (sczyh30) [inline sentinel-dashboard/src/main/webapp/resources/app/scripts/controllers/identity.js]: For the resource chain page, we might need to keep the v1 service. If users are going to push the rules to the config center, they could modify here.

File: sentinel-dashboard/src/main/webapp/resources/app/scripts/app.js [hunk 2]
```diff
@@ -155,6 +155,22 @@ angular
         }
       })
 
+      .state('dashboard.paramFlowV2', {
+        templateUrl: 'app/views/param_flow_v2.html',
+        url: '/v2/paramFlow/:app',
+        controller: 'ParamFlowControlleV2',
```
Comment (sczyh30) [inline sentinel-dashboard/src/main/webapp/resources/app/scripts/app.js]: typo: `ParamFlowControllerV2`

File: sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        entity.setId(null);
+        entity.getRule().setResource(entity.getResource().trim());
+        Date date = new Date();
+        entity.setGmtCreate(date);
+        entity.setGmtModified(date);
+        try {
+            entity = repository.save(entity);
+            publishRules(entity.getApp());
+            return Result.ofSuccess(entity);
+        } catch (ExecutionException ex) {
```
Comment (sczyh30) [inline sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java]: Just catch `Exception` (or `Throwable`) here is enough. You may refer to `FlowControllerV2`.

File: sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        entity.setId(id);
+        Date date = new Date();
+        entity.setGmtCreate(oldEntity.getGmtCreate());
+        entity.setGmtModified(date);
+        try {
+            entity = repository.save(entity);
+            publishRules(entity.getApp());
+            return Result.ofSuccess(entity);
+        } catch (ExecutionException ex) {
```
Comment (sczyh30) [inline sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java]: ditto

File: sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        if (port == null || port <= 0) {
+            return Result.ofFail(-1, "Invalid parameter: port");
+        }
+        if (!checkIfSupported(app, ip, port)) {
+            return unsupportedVersion();
+        }
+        try {
+            return Result.ofSuccess(ruleProvider.getRules(app));
+        } catch (ExecutionException ex) {
```
Comment (sczyh30) [inline sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java]: Just catch `Exception` (or `Throwable`) here is enough. You may refer to `FlowControllerV2`.

File: sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            // If error occurred or cannot retrieve machine info, return true.
+        } catch (Exception ex) {
+            return true;
+        }
+    }
+
+    @GetMapping("/rules")
+    public Result<List<ParamFlowRuleEntity>> apiQueryAllRulesForMachine(HttpServletRequest request,
+                                                                        @RequestParam String app,
+                                                                        @RequestParam String ip,
```
Comment (sczyh30) [inline sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java]: Here the `ip` and `port` parameter are not needed here. And we don't have to check versions here.

File: sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+            return Result.ofFail(-1, "ip cannot be null or empty");
+        }
+        if (port == null || port <= 0) {
+            return Result.ofFail(-1, "Invalid parameter: port");
+        }
+        if (!checkIfSupported(app, ip, port)) {
+            return unsupportedVersion();
+        }
+        try {
+            return Result.ofSuccess(ruleProvider.getRules(app));
```
Comment (sczyh30) [inline sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/controller/v2/ParamFlowRuleControllerV2.java]: Here we might need to add logic for handling cluster rules. You could refer to `FlowControllerV2` like this:

```java
 List<FlowRuleEntity> rules = ruleProvider.getRules(app);
            if (rules != null && !rules.isEmpty()) {
                for (FlowRuleEntity entity : rules) {
                    entity.setApp(app);
                    if (entity.getClusterConfig() != null && entity.getClusterConfig().getFlowId() != null) {
                        entity.setId(entity.getClusterConfig().getFlowId());
                    }
                }
            }
            rules = repository.saveAll(rules);
            return Result.ofSuccess(rules);
```

**Comment by sczyh30** [comment]
And the v2 page is intended for cluster-scope pushing, so no need to keep the machine list. Just list all rules from configuration center is okay :)

Besides, have you tested for publishing cluster parameter flow rules? Will it work correctly?

**Comment by threedr3am** [comment]
Think you for your correction.I will review this code and recode it.

**Pull request closed without merging**